Add GSoC 2025 post on UBSan improvements (#81)
llvm/llvm-blog-www: content/posts/2025-09-01-gsoc-ubsan-trap-messages.md

---
author: "Anthony Tran"
date: "2025-09-01"
tags: ["GSoC", "Clang", "CodeGen"]
title: "GSoC 2025: Usability Improvements for the Undefined Behavior Sanitizer"
---

## Introduction

My name is Anthony and I had the pleasure of working on improving the Undefined Behavior Sanitizer this Google Summer of Code 2025. My mentors were Dan Liew and Michael Buch.

## Background

Undefined Behavior Sanitizer (UBSan) is a tool for detecting a subset of the undefined behaviors in the C, C++, and Objective-C languages at runtime. This project focused mainly on the trapping variant of UBSan, which is evoked through `-fsanitize-trap=<...>` along with `-fsanitize=<...>`. Trapping UBSan is a lighter-weight version of UBSan because upon detection of undefined behavior a trap instruction is executed rather than calling into a runtime library to handle the undefined behavior. This makes it more appealing for kernel, embedded, and production hardening use cases.

> Here are some basic examples of undefined behavior:

Reviewer: Hmmm feels a bit weird to just have the source example without running them under UBSan. Should we, for each example, show UBSan complaining when running the program (not necessarily using trapping UBSan)

Author: Yes, I think removing the other examples would be fine. Dan had a good idea to just use one example and keep that as a running example for the entire blog post, which I think makes the most sense.

   3   
   4    int add(int a, int b) {
-> 5        return a + b;
   6    }
   7   
   8    int main(int argc, const char** argv) {
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BREAKPOINT (code=1, subcode=0x100003d3c)
  * frame #0: 0x0000000100003d3c add`add(a=2147483647, b=1) at add.c:5:13
    frame #1: 0x0000000100003eec add`main(argc=3, argv=0x000000016fdff110) at add.c:13:18
    frame #2: 0x00000001842bab98 dyld`start + 6076
(lldb) dis -p
add`add:
->  0x100003d3c <+40>: brk    #0x5500
    0x100003d40 <+44>: ldr    w0, [sp, #0x4]
    0x100003d44 <+48>: add    sp, sp, #0x10
    0x100003d48 <+52>: ret   
```

We can see that a `brk` instruction was hit while handling the `a + b` expression but there is no good explanation of what happened. `brk` is the trap instruction on arm64 but it is not particularly clear this has anything to do with UBSan. For this toy example we can speculate that integer overflow occurred because the program was built with trapping UBSan and the trap was hit while handling `a + b`. However, in real programs built with trapping UBSan and potentially other hardening mechanisms it is often far less obvious what happened.

> For this particular example. The information that this is an integer overflow UBSan check is actually there but it is not very obvious. On x86_64 and arm64 the reason for trapping is actually [encoded in the operand to the trap instruction](https://maskray.me/blog/2023-01-29-all-about-undefined-behavior-sanitizer ). In this case the `#0x5500` immediate to the brk instruction encodes that this is a UBSan trap for integer overflow. The UBSan immediate is encoded as `('U' << 8) + SanitizerHandler` where `SanitizerHandler` is the enum value from the `SanitizerHandler` enum inside Clang’s internals.

Reviewer: Should we provide a link to the [`SanitizerHandler` enum](https://github.com/llvm/llvm-project/blob/96195e7d44613e272475c90df187678036f21966/clang/lib/CodeGen/SanitizerHandler.h#L78)?

As we can see the debugging experience with UBSan traps is not ideal and improving this was the primary goal of the GSoC project.

## Human readable descriptions of UBSan traps in LLDB

During my GSoC project I implemented support for displaying human readable descriptions of UBSan traps in LLDB to improve the debugging experience.

> The approach used is based on how `__builtin_verbose_trap` is currently implemented inside Clang [11] [12]. `__builtin_verbose_trap` was implemented in the past for [libc++ hardening](https://discourse.llvm.org/t/rfc-hardening-in-libc/73925). At a high-level this works by encoding the reason for trapping as a string on the trap instruction in the debug info for the program being compiled. Then when a trap is hit in the debugger, the debugger retrieves this string and shows it as the reason for trapping.

Reviewer: ```suggestion
The approach used is based on how `__builtin_verbose_trap` is currently implemented inside Clang [11] [12]. `__builtin_verbose_trap` was implemented in the past for [libc++ hardening](https://discourse.llvm.org/t/rfc-hardening-in-libc/73925). At a high-level this works by encoding the reason for trapping on the trap's debug info. When a trap is hit in the debugger, the debugger retrieves this string and shows it as the reason for trapping.
```

Reviewer: Should we mention that you are going to expand on this in an upcoming section? Currently it sounds like you are glossing over the implementation details, although you *do* explain them later on

Reviewer: Actually, can we just remove this entire paragraph (or move the important bits into the `Encoding the trap reason in the debug info` subsection below?

I feel like gives us following clear structure:
1. "We want to make the trap reasons human readable in the debugger"
2. "We could make the debugger handle this without compiler support"
  * "These are the reasons it didn't work"
3. "Encode reason in debug info"
  * "This is the approach we took"

> ### An Alternative approach

Reviewer: ```suggestion
### Let debugger handle most of the work
```
Or something like that?

An alternative to this approach would be to teach debuggers (e.g. LLDB) to decode the trap reason encoded in trap instructions in the debugger. However, this approach wasn’t taken for several reasons:

> 

Reviewer: We need to make these a bullet point

Using the trap reason encoded in trap instructions only works for x86_64 and arm64. The approach that I used works for all targets where debug info is supported (many more).
Relying on decoding the trap reason encoded in the trap instruction creates a tight coupling between the compiler and the debugger because if the encoding ever changes
The debugger would need to be changed to adapt to the new encoding.
Older versions of the debugger would fail to work with binaries using the new encoding.
New versions of the debugger would fail to work with binaries using the old encoding.
In contrast, encoding the trap reason as a string in the debug info is a much looser coupling because the compiler is free to change the trap reason without changes to the debugger.

### Encoding the trap reason in the debug info

> As previously mentioned the approach I took is based on how `__builtin_verbose_trap` encodes its message into debug info. This is done by pretending in the debug info that the trap instruction was inlined from another function, where that function is artificially generated and its name is of the form `__clang_trap_msg$<Category>$<TrapMessage>`, where <Category> and <TrapMessage> are the trap category and message to display when trapping respectively. This function does not actually exist in the compiled program. It only exists in the debug info as a convenient (albeit hacky) way to describe the reason for trapping.

Reviewer: > This is done by pretending in the debug info that the trap instruction was inlined from another function

I wouldn't lead with this. I feel like this just opens a bunch of questions.

Can we just say that the trap reason string gets encoded on the trap's debug info. Then show the IR and point out that the trap's debug info metadata is scoped by the `DISubprogram` `%32`, which is not actually a real function, but we needed to stuff the string somewhere in the DWARF DIE tree and using a `DW_TAG_subprogram` was deemed the most straightforward (and most space efficient) place to put it. We could've put the string elsewhere but for reasons outside the scope of this blog pots it lives on a fake function DIE. The trap reason is encoded in the DW_TAG_subprograms name. You can point them to the discussion start at https://github.com/llvm/llvm-project/pull/145967#issuecomment-3054319138 for more info on this.

Using bloaty, I tested a release build of clang with the `-fsanitize-debug-trap-reasons` flag enabled, and one with it disabled (`-fno-sanitize-debug-trap-reasons`). We found that the size difference was negligible; results are below.

```
   FILE SIZE        VM SIZE   
--------------  --------------
 +0.3% +6.01Mi  +0.3% +6.01Mi    ,__debug_info
 +2.0% +2.26Mi  [ = ]       0    [Unmapped]
 +1.2% +1.35Mi  +1.2% +1.35Mi    ,__apple_names
 +0.0% +1.01Mi  +0.0% +1.01Mi    ,__debug_str
 +0.8%  +636Ki  +0.8%  +635Ki    ,__debug_line
 +0.4%  +161Ki  +0.4%  +161Ki    ,__debug_ranges
 +0.4% +47.9Ki  +0.4% +47.9Ki    ,__debug_abbrev
 +0.0%     +14  +0.0%     +14    ,__apple_types
 [ = ]       0  +0.0%      +8    ,__common
 [ = ]       0  +7.1%      +4    ,__thread_bss
 -0.0%      -4  -0.0%      -4    ,__const
 -0.0% -1.27Ki  -0.0% -1.27Ki    ,__cstring
 +0.2% +11.5Mi  +0.1% +9.19Mi    TOTAL
 ```

> Note it is likely the code size difference is negligible because because in optimized builds trap instructions in a function get merged together which causes the additional debug info my patch adds to be dropped.

Reviewer: I would point out where the debug info size impact would mainly come from: https://github.com/llvm/llvm-project/pull/145967#issuecomment-3068862442

And also mention, the size increase is tied to how many traps get emitted (since we emit a new DW_TAG_subprogram DIE for each trap). I would also link to the comment here: https://github.com/llvm/llvm-project/pull/154618#issuecomment-3225724300, which shows that in production Google saw a pretty large size increase in debug-info with this (~15% total binary size increase IIRC). So this could be a potential future thing to look into. Why does the binary size increase by that much and what could we do about it? As it stands, they had to disable the trap reason because of that, which isn't great.
